### ISSUE TITLE
Error in decoding job in WorkerManager causes Application crash

### ISSUE THREAD

**Issue by keyan** [edgurgel/verk#89]
We are seeing this issue come up:

```
15:49:13.960 [info]  10 jobs readded to the queue php_queue:emails from inprogress list

15:49:13.963 [error] Manager terminating, reason: {%Poison.SyntaxError{message: "Unexpected end of input", token: nil}, [{Poison.Parser, :parse!, 2, [file: 'lib/poison/parser.ex', line: 54]}, {Poison, :decode!, 2, [file: 'lib/poison.ex', line: 83]}, {Verk.Job, :decode!, 1, [file: 'lib/verk/job.ex', line: 19]}, {Verk.WorkersManager, :"-handle_info/2-fun-0-", 3, [file: 'lib/verk/workers_manager.ex', line: 102]}, {Enum, :"-reduce/3-lists^foldl/2-0-", 3, [file: 'lib/enum.ex', line: 1623]}, {Verk.WorkersManager, :handle_info, 2, [file: 'lib/verk/workers_manager.ex', line: 102]}, {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 601]}, {:gen_server, :handle_msg, 5, [file: 'gen_server.erl', line: 667]}]}

15:49:13.964 [error] GenServer :"php_queue:emails.workers_manager" terminating
** (Poison.SyntaxError) Unexpected end of input
    (poison) lib/poison/parser.ex:54: Poison.Parser.parse!/2
    (poison) lib/poison.ex:83: Poison.decode!/2
    (verk) lib/verk/job.ex:19: Verk.Job.decode!/1
    (verk) lib/verk/workers_manager.ex:102: anonymous fn/3 in Verk.WorkersManager.handle_info/2
    (elixir) lib/enum.ex:1623: Enum."-reduce/3-lists^foldl/2-0-"/3
    (verk) lib/verk/workers_manager.ex:102: Verk.WorkersManager.handle_info/2
    (stdlib) gen_server.erl:601: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:667: :gen_server.handle_msg/5
Last message: :timeout
State: %Verk.WorkersManager.State{monitors: :"php_queue:emails.workers_manager", pool_name: :"php_queue:emails.pool", pool_size: 10, queue_manager_name: :"php_queue:emails.queue_manager", queue_name: :"php_queue:emails", timeout: 1000}

15:49:13.966 [info]  Application <APPLICATION NAME> exited: shutdown
{"Kernel pid terminated",application_controller,"{application_terminated,<APPLICATION NAME>,shutdown}"}
Kernel pid terminated (application_controller) ({application_terminated,<APPLICATION NAME>,shutdown})

Crash dump is being written to: erl_crash.dump...done
```

It looks like what is happening is that a [WorkerManager](https://github.com/edgurgel/verk/blob/master/lib/verk/workers_manager.ex#L102) attempts to decode a job that probably has bad JSON, this makes the worker manager crash and eventually leads to the parent process crashing.

I am confused as to why this would result in our application crashing. We are supervising Verk in the suggested way, I would expect instead for the worker manager process to crash and be restarted by `Verk.Supervisor`. Here is our supervisor configuration:

```
import Supervisor.Spec, warn: false
tree = [supervisor(Verk.Supervisor, [])]
opts = [name: <APPLICATION_NAME>.Supervisor, strategy: :one_for_one]
Supervisor.start_link(tree, opts)
```

So:

1. Why is our application crashing?
2. Is this expected?
3. What if we call `Poison.decode` instead and then move jobs with bad JSON into a `malformed` key in Redis?
    - We can't mark them as `dead` through `DeadSet` without modifications, because we [rely on proper encoding](https://github.com/edgurgel/verk/blob/master/lib/verk/dead_set.ex#L23) in `DeadSet.add/3`.

**Comment by edgurgel** [comment]
> Why is our application crashing?
> Is this expected?

The application dies cause the Verk.Supervisor is restarting too many times (because the workers manager is restarting too many times). It's really simple to tweak the restart strategy to be something like: "max restart times 99999" and possibly `:infinity` so that `Verk.Supervisor` will never cause a restart of the supervisor above (the App supervisor probably)

We could also discuss that the workers manager should never give up restarting (but then what if there's actually a bug on it or something?)

> What if we call Poison.decode instead and then move jobs with bad JSON into a malformed key in Redis?

That's a hard one. I'm not sure what's the best approach when the data can't be parsed properly. This should be an edge case, right? I would like some solution that doesn't complicate a lot the flow of data but it also gives enough information to the user that something was wrong. Maybe logging for now would be enough?

**Comment by mitchellhenke** [comment]
Would it be possible to add a new GenEvent and report it that way along with logging?

**Comment by keyan** [comment]
Thank you both for providing some help and suggestions! To answer them:

>We could also discuss that the workers manager should never give up restarting (but then what if there's actually a bug on it or something?)

I agree, this would not be a good thing to do. I'm fine with not changing that.

> I'm not sure what's the best approach when the data can't be parsed properly. This should be an edge case, right?

Possibly? Our usage right now is to encode the jobs into JSON from a legacy PHP app and write to Redis, Verk then picks up the job. I imagine this is likely a common workflow, although maybe coming from Ruby applications.

> I would like some solution that doesn't complicate a lot the flow of data but it also gives enough information to the user that something was wrong.

I agree that a complication of data movement is not ideal. I did some initial work, but admittedly it does complicate things: https://github.com/edgurgel/verk/compare/master...keyan:malformed_set_encoding_issues?expand=1

> Maybe logging for now would be enough?

Are you suggesting we delete the job and log it? I think that would be reasonable in general, but in my case I have ~5000 emails stuck in Redis, probably a large number are malformed. I probably have to recover the data somehow if Verk cannot process it, so logging and deleting the job wouldn't be enough.

----

> Would it be possible to add a new GenEvent and report it that way along with logging?

@mitchellhenke, as in do the logging inside of a handler? Is this what you mean?

```elixir
def handle_event({:malformed, job}, message) do
  Logger.error "#{message}"
end
```

----

I agree we have to log something. What are your thoughts on finding a way to put the jobs into the `DeadSet`? I think it makes sense that malformed jobs would be sent there, it doesn't complicate the flow too much, and still allows for a way to keep the jobs should you need to recover them.

Thoughts?

**Comment by keyan** [comment]
Okay, I poked around and tried implementing something to send jobs to the dead queue when malformed, and it was pretty messy. I'm okay with just logging the job for now.

Any suggestions on how to handle this? `QueueManager.dequeue` will have already moved the job into the `inprogress` queue, see [here](https://github.com/edgurgel/verk/blob/b9db13cd5afa008ccc51c85fb3c8c807dc18bea3/lib/verk/workers_manager.ex#L100-L102). We need to remove the job before logging and skipping it. Any input is appreciated!

**Comment by edgurgel** [comment]
Yeah we need to add to the dead set and then remove from the `inprogress` acknowledging the job.

**Comment by keyan** [comment]
@edgurgel, oh okay. I misunderstood. I thought you didn't want to move it to the dead set, just log and remove from `inprogress` (which is what I did in #91). I'll open a new PR that instead sends to the dead set.

**Comment by edgurgel** [comment]
Yeah I wasn't clear! I'm not super decided on which solution to take but logging is way easier :D

**Comment by keyan** [comment]
Haha no worries. I agree, logging is way easier. I worked on first pass which sent the jobs to the dead set and it complicated things more than I'd like. Given that you are on the fence and the complication you mentioned in https://github.com/edgurgel/verk/pull/91#discussion_r86055106, I am going to go forward with just logging and removing the job.

I ran the changes in #91 in production and we were able to clear our queues, only 5 jobs had the malformed data issue. I think this is an edge case as you mentioned and probably not worth muddying up dead set because of it. If we decide to eventually send these to the dead set, the change will be add on #91 easily.
